File: app/src/main/java/com/dmrandevu/gallery/data/SettingsStore.kt
package com.dmrandevu.gallery.data

import android.content.SharedPreferences
import androidx.core.content.edit

/** Remembers everything needed to log back in except the password. */
class SettingsStore(private val prefs: SharedPreferences) {

    var baseUrl: String
        get() = prefs.getString(KEY_BASE_URL, DEFAULT_BASE_URL)!!
        set(value) = prefs.edit { putString(KEY_BASE_URL, value.trimEnd('/')) }

    var adminUsername: String
        get() = prefs.getString(KEY_ADMIN, "")!!
        set(value) = prefs.edit { putString(KEY_ADMIN, value.trim()) }

    var igUsername: String
        get() = prefs.getString(KEY_IG, DEFAULT_IG_ACCOUNT)!!
        set(value) = prefs.edit { putString(KEY_IG, value.trim().removePrefix("@")) }

    /** Blur faces in every exported video. Off by default: it re-encodes, which takes a while. */
    var blurFaces: Boolean
        get() = prefs.getBoolean(KEY_BLUR_FACES, DEFAULT_BLUR_FACES)
        set(value) = prefs.edit { putBoolean(KEY_BLUR_FACES, value) }

    /** Blur licence plates in every exported video. Off by default, like the face filter. */
    var blurPlates: Boolean
        get() = prefs.getBoolean(KEY_BLUR_PLATES, DEFAULT_BLUR_PLATES)
        set(value) = prefs.edit { putBoolean(KEY_BLUR_PLATES, value) }

    /**
     * Run the plate detector at the smaller input size: about 40% quicker, and it finds roughly
     * two thirds as many plates. On by default while the trade is being lived with.
     */
    var fastPlates: Boolean
        get() = prefs.getBoolean(KEY_FAST_PLATES, DEFAULT_FAST_PLATES)
        set(value) = prefs.edit { putBoolean(KEY_FAST_PLATES, value) }

    /**
     * Beep over Turkish swearing in every exported video. Off by default: it needs a third of a
     * gigabyte of models downloaded before it can do anything.
     */
    var censorAudio: Boolean
        get() = prefs.getBoolean(KEY_CENSOR_AUDIO, DEFAULT_CENSOR_AUDIO)
        set(value) = prefs.edit { putBoolean(KEY_CENSOR_AUDIO, value) }

    /**
     * Beep milder insults too, not only outright profanity. Off by default — the spike had
     * "manyak" firing on a clip where nobody swore.
     */
    var censorInsults: Boolean
        get() = prefs.getBoolean(KEY_CENSOR_INSULTS, DEFAULT_CENSOR_INSULTS)
        set(value) = prefs.edit { putBoolean(KEY_CENSOR_INSULTS, value) }

    /** Drift the account handle across every exported video, so a repost still shows whose it is. */
    var watermark: Boolean
        get() = prefs.getBoolean(KEY_WATERMARK, DEFAULT_WATERMARK)
        set(value) = prefs.edit { putBoolean(KEY_WATERMARK, value) }

    companion object {
        const val DEFAULT_BASE_URL = "https://dmrandevu.com"
        const val DEFAULT_IG_ACCOUNT = "trafik_cezasi"
        const val DEFAULT_BLUR_FACES = false
        const val DEFAULT_BLUR_PLATES = false
        const val DEFAULT_FAST_PLATES = true
        const val DEFAULT_WATERMARK = false
        const val DEFAULT_CENSOR_AUDIO = false
        const val DEFAULT_CENSOR_INSULTS = false
        private const val KEY_BASE_URL = "base_url"
        private const val KEY_ADMIN = "admin_username"
        private const val KEY_IG = "ig_username"
        private const val KEY_BLUR_FACES = "blur_faces"
        private const val KEY_BLUR_PLATES = "blur_plates"
        private const val KEY_FAST_PLATES = "fast_plates"
        private const val KEY_WATERMARK = "watermark"
        private const val KEY_CENSOR_AUDIO = "censor_audio"
        private const val KEY_CENSOR_INSULTS = "censor_insults"
    }
}

File: ios/DMRandevuGaleri/Data/SettingsStore.swift
import Foundation

/// Remembers everything needed to log back in except the password.
final class SettingsStore {

    static let defaultBaseURL = "https://dmrandevu.com"
    static let defaultIhbarBaseURL = "https://ihbar.lega.digital"
    static let defaultIGAccount = "trafik_cezasi"

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // UserDefaults answers `false` for an absent Bool, which happens to be the wanted
        // default for all three filters — but not for the plate speed, which starts on.
        defaults.register(defaults: [Key.fastPlates: true])
    }

    var baseURL: String {
        get { defaults.string(forKey: Key.baseURL) ?? Self.defaultBaseURL }
        set {
            var trimmed = newValue
            while trimmed.hasSuffix("/") { trimmed.removeLast() }
            defaults.set(trimmed, forKey: Key.baseURL)
        }
    }

    var adminUsername: String {
        get { defaults.string(forKey: Key.admin) ?? "" }
        set { defaults.set(newValue.trimmingCharacters(in: .whitespaces), forKey: Key.admin) }
    }

    var igUsername: String {
        get { defaults.string(forKey: Key.ig) ?? Self.defaultIGAccount }
        set {
            let handle = newValue.trimmingCharacters(in: .whitespaces)
            defaults.set(handle.hasPrefix("@") ? String(handle.dropFirst()) : handle, forKey: Key.ig)
        }
    }

    /// Trafik İhbar sunucusunun adresi.
    ///
    /// DMRandevu'dan AYRI bir sistem ve ayrı bir sunucu; ``baseURL`` ile
    /// karıştırılmamalı. Ayrı durmasının sebebi yalnızca düzen değil: galeri
    /// sunucusu değiştiğinde (yerel bir kopyaya bakarken) ihbar sunucusunun
    /// onunla birlikte kaymaması gerekiyor.
    var ihbarBaseURL: String {
        get { defaults.string(forKey: Key.ihbarBaseURL) ?? Self.defaultIhbarBaseURL }
        set {
            var trimmed = newValue.trimmingCharacters(in: .whitespaces)
            while trimmed.hasSuffix("/") { trimmed.removeLast() }
            defaults.set(trimmed, forKey: Key.ihbarBaseURL)
        }
    }

    /// İhbar sunucusunun cihaz belirteci ("tid_…").
    ///
    /// Yönetici konsolunda bir kez üretiliyor ve bir daha gösterilmiyor, o yüzden
    /// buraya bir kez yapıştırılıp saklanıyor. Boşken ihlal düğmesi ağa hiç
    /// çıkmıyor ve bunu kullanıcıya söylüyor.
    var ihbarToken: String {
        get { defaults.string(forKey: Key.ihbarToken) ?? "" }
        set { defaults.set(newValue.trimmingCharacters(in: .whitespaces), forKey: Key.ihbarToken) }
    }

    /// Blur faces in every exported video. Off by default: it re-encodes, which takes a while.
    var blurFaces: Bool {
        get { defaults.bool(forKey: Key.blurFaces) }
        set { defaults.set(newValue, forKey: Key.blurFaces) }
    }

    /// Blur licence plates in every exported video. Off by default, like the face filter.
    var blurPlates: Bool {
        get { defaults.bool(forKey: Key.blurPlates) }
        set { defaults.set(newValue, forKey: Key.blurPlates) }
    }

    /// Run the plate detector at the smaller input size: quicker, and it finds fewer of the
    /// smaller plates. On by default while the trade is being lived with.
    var fastPlates: Bool {
        get { defaults.bool(forKey: Key.fastPlates) }
        set { defaults.set(newValue, forKey: Key.fastPlates) }
    }

    /// Drift the account handle across every exported video, so a repost still shows whose it is.
    var watermark: Bool {
        get { defaults.bool(forKey: Key.watermark) }
        set { defaults.set(newValue, forKey: Key.watermark) }
    }

    /// Akış tanıtımının en son gösterildiği yapı ("1.0+1"); boş ise hiç gösterilmedi.
    ///
    /// BAYRAK DEĞİL YAPI ADI: hareketler değiştiğinde yapı numarası artırılarak tanıtım bir kez
    /// daha gösterilebiliyor. Düz bir Bool olsaydı, değişen bir hareketi öğrenmenin tek yolu
    /// uygulamayı silip yeniden kurmak olurdu.
    var tourShownBuild: String {
        get { defaults.string(forKey: Key.tourBuild) ?? "" }
        set { defaults.set(newValue, forKey: Key.tourBuild) }
    }

    /// Beep over Turkish swearing in every exported video. Off by default: it needs a quarter of
    /// a gigabyte of models downloaded before it can do anything.
    var censorAudio: Bool {
        get { defaults.bool(forKey: Key.censorAudio) }
        set { defaults.set(newValue, forKey: Key.censorAudio) }
    }

    /// Whether the censor listens to the video or only beeps what was marked by hand.
    ///
    /// Automatic on a clip the recognizer manages; by hand on the ones it does not, where running
    /// it costs minutes to be told what the operator has already said.
    var censorByHand: Bool {
        get { defaults.bool(forKey: Key.censorByHand) }
        set { defaults.set(newValue, forKey: Key.censorByHand) }
    }

    /// Beep milder insults too, not only outright profanity. Off by default — "manyak" fired on a
    /// clip where nobody swore.
    var censorInsults: Bool {
        get { defaults.bool(forKey: Key.censorInsults) }
        set { defaults.set(newValue, forKey: Key.censorInsults) }
    }

    private enum Key {
        static let baseURL = "base_url"
        static let ihbarBaseURL = "ihbar_base_url"
        static let ihbarToken = "ihbar_token"
        static let admin = "admin_username"
        static let ig = "ig_username"
        static let blurFaces = "blur_faces"
        static let blurPlates = "blur_plates"
        static let fastPlates = "fast_plates"
        static let watermark = "watermark"
        static let censorAudio = "censor_audio"
        static let censorInsults = "censor_insults"
        static let censorByHand = "censor_by_hand"
        static let tourBuild = "tour_shown_build"
    }
}
